### ISSUE TITLE
Invalid function pointer '0' called with signature 'v' happens with seemingly valid code

### ISSUE THREAD

**Issue by fictitious** [kripken/emscripten#6532]
with these 3 files [available as gist](https://gist.github.com/fictitious/a2548e8699a7b51f58d3d1972eb38e42), also inlined here, I'm getting this exception at runtime:

Invalid function pointer '0' called with signature 'v'. Perhaps this is an invalid value 
(e.g. caused by calling a virtual method on a NULL pointer)? Or calling a function with
an incorrect type, which will fail?  (it is worth building your source files with -Werror
(warnings are errors), as warnings can indicate undefined behavior which can cause this)
This pointer might make sense in another type signature: vi: 0  vii: 0  viii: 0  viiii: 0  viiiii: 0  viiiiii: 0  viiiiiii: 0  viiiiiiiiii: 0  viiiiiiiiiiiiiii: 0  i: 0  ii: 0  iii: 0  diii: 0  iiii: 0  iiiii: 0  iiiiid: 0  iiiiii: 0  iiiiiid: 0  iiiiii: 0  iiiiiiii: 0  iiiiiiiii: 0  iiiiiiiiiii: 0  iiiiiiiiiiii: 0  iiiiiiiiiiiii: 0  

This is reduced example from some C++ code which used javascript callbacks. It worked fine until I made js callbacks truly async - then it started throwing assertions. Any help is much appreciated.

emcc -v:
(emscripten-fastcomp built from source, ubuntu v16.04)

```
emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 1.37.38
clang version 5.0.0 (https://github.com/kripken/emscripten-fastcomp-clang 73beffb52d95ec01d2d3ce707439d9cb807c7bc2) (https://github.com/kripken/emscripten-fastcomp eea5a74d8f8a9838cf79d9f3b78d39a8db3f6422) (emscripten 1.37.38 : 1.37.38)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /home/..../emscripten-fastcomp/build/bin
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/5
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/5.4.0
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/6
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/6.0.0
Selected GCC installation: /usr/lib/gcc/x86_64-linux-gnu/5.4.0
Candidate multilib: .;@m64
Selected multilib: .;@m64

```

c++ file: async_callback_test.cpp

```c++
#include <iostream>
#include <emscripten.h>

extern "C" {

  void call_async_operation() {

    EM_ASM({AsyncOperation.start()});


    while (!EM_ASM_INT({return AsyncOperation.done})) {
      emscripten_sleep(200);
    }
  }


  // remove throw() => no crash
  static void nothrow_func() throw()
  {
    call_async_operation();

    // remove the following line => no crash
    std::cout << "async operation OK" << std::endl;
  }

  void async_callback_test() {
    nothrow_func();
  }

}
```
javascript file: use.js

```javascript

const AsyncOperation = {
    done: false,

    start() {
        // this.done = true; // uncomment this line => no crash
        Promise.resolve().then(() => this.done = true);
    }
};


Module.ccall('async_callback_test', null, [], [], { async: true });
```

build and run script: run.sh

```bash
mkdir -p out

em++ -Wall -O0 -std=c++0x -fPIC -fPIC -c -o out/async_callback_test.o async_callback_test.cpp

emar rcvs out/libasync_callback_test.a  out/async_callback_test.o

emcc out/libasync_callback_test.a -o out/libasync_callback_test.js \
	-O0 \
	-s EXPORTED_FUNCTIONS="['_async_callback_test']" \
	-s EXTRA_EXPORTED_RUNTIME_METHODS='["ccall"]' \
	-s DISABLE_EXCEPTION_CATCHING=0 \
	-s ALLOW_MEMORY_GROWTH=1 \
	-s EMTERPRETIFY=1 \
	-s EMTERPRETIFY_ASYNC=1 \
	-s ASSERTIONS=2 \
	-s SAFE_HEAP=1 \
	-s DEMANGLE_SUPPORT=1 \
	--profiling-funcs \
	--minify 0 \
	--memory-init-file 0


cat out/libasync_callback_test.js use.js > out/run.js

node out/run.js

```

**Comment by kripken** [comment]
Thanks for the reduced testcase! This turned out to be a subtle bug combining the emterpreter, async, SAFE_HEAP, and invokes. Fix and details in #6539.

**Comment by fictitious** [comment]
Thanks! Did not know that it was caused by SAFE_HEAP, not using it solves my problem!
